discovery_engine/lib/src/ffi/types/engine.dart

```diff
@@ -0,0 +1,187 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import 'dart:ffi' show nullptr, Pointer;
+import 'dart:typed_data' show Float32List, Uint8List;
+
+import 'package:xayn_discovery_engine/src/domain/models/configuration.dart'
+    show Configuration;
+import 'package:xayn_discovery_engine/src/domain/models/document.dart'
+    show UserReaction;
+import 'package:xayn_discovery_engine/src/domain/models/feed_market.dart'
+    show FeedMarket;
+import 'package:xayn_discovery_engine/src/domain/models/unique_id.dart'
+    show DocumentId, StackId;
+import 'package:xayn_discovery_engine/src/ffi/genesis.ffigen.dart'
+    show RustResultSharedEngineString, RustSharedEngine;
+import 'package:xayn_discovery_engine/src/ffi/load_lib.dart'
+    show asyncCore, ffi;
+import 'package:xayn_discovery_engine/src/ffi/types/box.dart' show Boxed;
+import 'package:xayn_discovery_engine/src/ffi/types/document/document.dart'
+    show DocumentFfi;
+import 'package:xayn_discovery_engine/src/ffi/types/document/time_spent.dart'
+    show TimeSpentFfi;
+import 'package:xayn_discovery_engine/src/ffi/types/document/user_reacted.dart'
+    show UserReactedFfi;
+import 'package:xayn_discovery_engine/src/ffi/types/feed_market_vec.dart'
+    show FeedMarketSliceFfi;
+import 'package:xayn_discovery_engine/src/ffi/types/init_config.dart'
+    show InitConfigFfi;
+import 'package:xayn_discovery_engine/src/ffi/types/primitives.dart'
+    show Uint8ListFfi;
+import 'package:xayn_discovery_engine/src/ffi/types/result.dart'
+    show
+        resultSharedEngineStringFfiAdapter,
+        resultVecDocumentStringFfiAdapter,
+        resultVecU8StringFfiAdapter,
+        resultVoidStringFfiAdapter;
+import 'package:xayn_discovery_engine/src/infrastructure/assets/native/data_provider.dart'
+    show NativeSetupData;
+
+/// A handle to the discovery engine.
+class DiscoveryEngineFfi {
```

Comment:
This should probably implement `Engine` once [TY-2392](https://xainag.atlassian.net/browse/TY-2392), [TY-2393](https://xainag.atlassian.net/browse/TY-2393) , [TY-2397](https://xainag.atlassian.net/browse/TY-2397) land.

Through engine has currently following `Map<Document, ActiveDocumentData> getFeedDocuments(int maxDocuments);` which I believe needs to be changed to `List<Document> getFeedDocuments...` as the place calling the engine  (the `DocumentManager`) has full access to `ActiveDocumentData` but not the engine itself.

Reply:
i thought that the `DiscoveryEngineFfi` will be wrapped/extended by some `DiscoveryEngine` later on, but this would also be possible, let's align in the daily.